dkulib/parallelizer/parallelizer.py

```diff
@@ -0,0 +1,254 @@
+        batch_support: If True, send batches of row (list of dict) to the `function`
+            Else (default) send rows as dict to the function
+        batch_size: Number of rows to include in each batch. Default is 10.
+            Taken into account if `batch_support` is True.
+        batch_response_parser: Function used to parse the raw response from the function in batch mode,
+            and assign the actual responses and errors back to the original batch of row (list of dict).
+            This is often required for batch APIs which return nested objects with a mix of responses and errors.
+        output_column_prefix: Column prefix to add to the output columns of the dataframe,
+            containing the `function` responses and errors. Default is "output".
+        verbose: If True, log additional information on errors.
+            Else (default) log the error message and the error type.
+
+    """
+
+    DEFAULT_PARALLEL_WORKERS = 4
+    DEFAULT_BATCH_SIZE = 10
+    DEFAULT_BATCH_SUPPORT = False
+    DEFAULT_VERBOSE = False
+    DEFAULT_OUTPUT_COLUMN_PREFIX = "output"
+    OUTPUT_COLUMN_NAME_DESCRIPTIONS = OrderedDict(
+        [
+            ("response", "Raw response in JSON format"),
+            ("error_message", "Error message"),
+            ("error_type", "Error type or code"),
+            ("error_raw", "Raw error"),
+        ]
+    )
+    """Default dictionary of output column names (key) and their descriptions (value)"""
+
+    def __init__(
+        self,
+        function: Callable[[Union[Dict, List[Dict]]], Union[Dict, List[Dict]]],
+        error_handling: ErrorHandling = ErrorHandling.LOG,
+        exceptions_to_catch: Tuple[Exception] = (),
+        parallel_workers: int = DEFAULT_PARALLEL_WORKERS,
+        batch_support: bool = DEFAULT_BATCH_SUPPORT,
+        batch_size: int = DEFAULT_BATCH_SIZE,
+        batch_response_parser: Optional[Callable[[List[Dict], Any, NamedTuple], List[Dict]]] = None,
+        output_column_prefix: AnyStr = DEFAULT_OUTPUT_COLUMN_PREFIX,
+        verbose: bool = DEFAULT_VERBOSE,
+    ):
+        self.function = function
+        self.error_handling = error_handling
+        self.exceptions_to_catch = exceptions_to_catch
+        if error_handling == ErrorHandling.LOG and not exceptions_to_catch:
+            raise ValueError("Please set at least one exception in exceptions_to_catch")
+        self.parallel_workers = parallel_workers
+        self.batch_support = batch_support
+        self.batch_size = batch_size
+        self.batch_response_parser = batch_response_parser
+        if batch_support and not batch_response_parser:
+            raise ValueError("Please provide a valid batch_response_parser function")
+        self.output_column_prefix = output_column_prefix
+        self.verbose = verbose
+
+    def _get_unique_output_column_names(self, existing_names: List[AnyStr]) -> NamedTuple:
+        """Return a named tuple with prefixed column names and their descriptions"""
```

Comment:
Good remark. The return of this function is indeed a NamedTuple with always the same names, but values that vary depending on the input dataframe (fed to `existing_names`).

I agree that this function is not very straightforward, but I am not sure how to change it. Do you have a suggestion?

Reply:
I have made output_column_names into a class private attribute in 475b191.

However, I can't really rename the columns at the end. That would create a risk of creating duplicate keys/columns. 

I can't easily rely on column numerical indexing because I have to translate to/from dictionaries and dataframes. IMO `output_column_names.response` and `output_column_names.error_message` are more readable than `output_column_names[0]` and `output_column_names[1]`.